### PR TITLE
chore: removes BOM

### LINKED ISSUE #6909: Remove UTF-8 BOM
### Introduce the problem

I encountered some issues while building the documentation. And I found that it was due to some files in the repository containing UTF-8 BOM. Maybe we should fix them. Here is a list:

```shell
$ git --no-pager grep --perl-regexp -I -l -e '^\xEF\xBB\xBF'
.devcontainer/__lv_conf.h__
demos/render/lv_demo_render.c
demos/render/lv_demo_render.h
env_support/cmsis-pack/lv_conf_cmsis.h
examples/styles/lv_example_style.h
examples/styles/lv_example_style_16.c
examples/styles/lv_example_style_17.c
examples/styles/lv_example_style_18.c
lv_conf_template.h
src/draw/sw/blend/lv_draw_sw_blend.c
src/draw/sw/blend/lv_draw_sw_blend_to_al88.c
src/draw/sw/blend/lv_draw_sw_blend_to_al88.h
src/draw/sw/blend/lv_draw_sw_blend_to_argb8888.c
src/draw/sw/blend/lv_draw_sw_blend_to_l8.c
src/draw/sw/blend/lv_draw_sw_blend_to_l8.h
src/draw/sw/blend/lv_draw_sw_blend_to_rgb565.c
src/draw/sw/blend/lv_draw_sw_blend_to_rgb888.c
src/draw/sw/lv_draw_sw.h
src/draw/sw/lv_draw_sw_fill.c
src/draw/sw/lv_draw_sw_gradient.c
src/draw/sw/lv_draw_sw_gradient.h
src/draw/sw/lv_draw_sw_img.c
src/draw/sw/lv_draw_sw_transform.c
src/drivers/windows/lv_windows_context.c
src/drivers/windows/lv_windows_context.h
src/drivers/windows/lv_windows_input.c
src/libs/thorvg/tvgSwRasterAvx.h
src/misc/lv_color.c
src/misc/lv_color.h
src/misc/lv_math.c
src/misc/lv_math.h
src/misc/lv_style.h
src/osal/lv_windows.c
src/osal/lv_windows.h
src/stdlib/builtin/lv_tlsf.h
src/stdlib/lv_mem.h
src/widgets/canvas/lv_canvas.c
tests/unity/unity_support.c
```

### Proposal

This problem has occurred before (#4684, #4685). Perhaps we can configure CI to automatically check for this to avoid having to deal with it repeatedly.

### PR DESCRIPTION
**Pull request by kdschlosser**
Fixes #6909

### CODE REVIEW

**Comment by Narukara** [comment]
Uh, there's already a PR #6942 to fix this.

**Comment by kdschlosser** [comment]
I saw that after I made this PR, 

This PR also adds the CI testing for UTF-8 BOM UTF-16LE BOM and UTF-16BE BOM.

**Comment by kdschlosser** [comment]
It also removes the BOM signature used in a comment line.

**Comment by kdschlosser** [comment]
@kisvegabor 

Let me know how you want to handle the BOM PR's. IO didn't realize you had made one. This PR includes adding a CI test for the 3 different types of BOM's used and it also removes the BOM signature used in a comment for the Persian font

I can revert the changes made to the files you already made the changes to or you can use this PR. it doesn't matter to me how you want to go about it.

File: src/font/lv_font_dejavu_16_persian_hebrew.c [hunk 1]
```diff
@@ -5860,7 +5860,7 @@ static LV_ATTRIBUTE_LARGE_CONST const uint8_t glyph_bitmap[] = {
     0xdd, 0x0, 0x8, 0x7c, 0xf7, 0x3f, 0xa4, 0xb,
     0xfc, 0x40, 0x7, 0xfb,
 
-    /* U+FEFF "﻿" */
+    /* U+FEFF "" */
```
Comment (liamHowatt) [inline src/font/lv_font_dejavu_16_persian_hebrew.c]: > comment in font file had BOM signature. Not sure why the comment was even there.

I think it just happens to share the same byte sequence as a UTF-16BE BOM signature but it's a valid unicode character.

BOM signatures can only come at the beginning. Could you update the CI to only check the beginning of the file?

File: .github/workflows/check_bom.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+          fetch-depth: 0
+
+      - name: UTF-8 BOM
+        run: |          
+          if grep -lrE --include="*.c" $'\xEF\xBB\xBF' .; then exit 1; else exit 0; fi
+          if grep -lrE --include="*.h" $'\xEF\xBB\xBF' .; then exit 1; else exit 0; fi
```
Comment (liamHowatt) [inline .github/workflows/check_bom.yml]: I tested and the CI only catches `.c` files. If all the C files are okay, it runs `exit 0`, skipping the rest of the steps.

File: src/font/lv_font_dejavu_16_persian_hebrew.c (hunk 1 above)
Comment (kdschlosser) [reply src/font/lv_font_dejavu_16_persian_hebrew.c]: I am not all that proficient in bash/sh scripting and I am not sure how I would go about having it check only the first couple of characters in the files. Maybe someone might be able to shed some light on that.

Out of question, what is the reasoning for using UTF-16 BOM as a font character?

File: .github/workflows/check_bom.yml (hunk 2 above)
Comment (kdschlosser) [reply .github/workflows/check_bom.yml]: Ohhhh.... I know what is happening. The CI is interpreting each exit call as a literal exit from the CI and not an exit from the script. that is why the next lines don't run.


Just need to remove the `else exit 0;` from each line.

Comment (kdschlosser) [reply .github/workflows/check_bom.yml]: OK this issue should be fixed now..

File: src/font/lv_font_dejavu_16_persian_hebrew.c (hunk 1 above)
Comment (kdschlosser) [reply src/font/lv_font_dejavu_16_persian_hebrew.c]: This should also be fixed now...

**Review by liamHowatt — COMMENTED** [review]
Seems to be working well now, as far as I can tell.

Can you revert [4a3dac4](https://github.com/lvgl/lvgl/pull/6961/commits/4a3dac4425289befdfa39a54910860f034a44112)?

> Out of question, what is the reasoning for using UTF-16 BOM as a font character?

That sequence is a unicode codepoint encoded as UTF-8 and that's the only thing it can possibly mean in a UTF-8 file such as that. In a "UTF-16 with BOM" file, it would be invalid to have that sequence, but the file is UTF-8, not "UTF-16 with BOM", so it's valid.

File: .github/workflows/check_bom.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+          grep -oarlE -m 1 --include="*.rst" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+          grep -oarlE -m 1 --include="*.txt" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+          grep -oarlE -m 1 --include="*.mk" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+          grep -oarlE -m 1 --include="*.py" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+          grep -oarlE -m 1 --include="Makefile" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+          grep -oarlE -m 1 --include="*.cmake" $'\xEF\xBB\xBF' . | while read -r line; do a=$(head -c 3 $line); if [ "${a}" = $'\xEF\xBB\xBF' ]; then echo "${line}" && exit 1; fi; done
+
```
Comment (kisvegabor) [inline .github/workflows/check_bom.yml:26]: Consider this:
```
grep -oarlE -m 1 --include="*.{c,h,rst,txt,mk,py,cmake}" --include="Makefile" $'\xEF\xBB\xBF' . | while read -r line; do
  a=$(head -c 3 "$line")
  if [ "${a}" = $'\xEF\xBB\xBF' ]; then
    echo "${line}" && exit 1
  fi
done
```

**Comment by kisvegabor** [comment]
@kdschlosser Do you plan to finish it? If so please consider [this](https://github.com/lvgl/lvgl/pull/6961#discussion_r1802887535)

**Comment by kdschlosser** [comment]
I thought this was finished. I had made the changes to fix it not running properly and I also updated it so it checks only the first few bytes for the BOM. is there something else that needs to be done?

**Comment by kdschlosser** [comment]
have you tried that code to make sure it works? :smile: and I just noticed I need to revert that single commit that changes the font file.

**Comment by kdschlosser** [comment]
OK I merged in the master branch changes, folved the conflicts, reverted that commit that alters the font file and made the changes @kisvegabor suggested.

File: .github/workflows/check_bom.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          grep -oarlE -m 1 --include="*.{c,h,rst,txt,mk,py,cmake}" --include="Makefile" $'\xFF\xFE' . | while read -r line; do
+            a=$(head -c 3 "$line")
+            if [ "${a}" = $'\xFF\xFE' ]; then
+              echo "${line}" && exit 1
+            fi
+          done
+
+      - name: UTF-16BE BOM
+        run: |
+          grep -oarlE -m 1 --include="*.{c,h,rst,txt,mk,py,cmake}" --include="Makefile" $'\xFE\xFF' . | while read -r line; do
```
Comment (liamHowatt) [inline .github/workflows/check_bom.yml]: I tested it in my own branch and unfortunately it did not catch a BOM until I changed the `--include`. Seems like the GLOB using `{}` isn't valid.

```suggestion
        run: |
          grep -oarlE -m 1 --include="*.c" --include="*.h" --include="*.rst" --include="*.txt" --include="*.mk" --include="*.py" --include="*.cmake" --include="Makefile" $'\xEF\xBB\xBF' . | while read -r line; do
            a=$(head -c 3 "$line")
            if [ "${a}" = $'\xEF\xBB\xBF' ]; then
              echo "${line}" && exit 1
            fi
          done

      - name: UTF-16LE BOM
        run: |
          grep -oarlE -m 1 --include="*.c" --include="*.h" --include="*.rst" --include="*.txt" --include="*.mk" --include="*.py" --include="*.cmake" --include="Makefile" $'\xFF\xFE' . | while read -r line; do
            a=$(head -c 3 "$line")
            if [ "${a}" = $'\xFF\xFE' ]; then
              echo "${line}" && exit 1
            fi
          done

      - name: UTF-16BE BOM
        run: |
          grep -oarlE -m 1 --include="*.c" --include="*.h" --include="*.rst" --include="*.txt" --include="*.mk" --include="*.py" --include="*.cmake" --include="Makefile" $'\xFE\xFF' . | while read -r line; do
```

Comment (kdschlosser) [reply .github/workflows/check_bom.yml]: I was just doing as I was instructed. I know the previous code did work so I will revert the commit for this change.

Comment (kisvegabor) [reply .github/workflows/check_bom.yml]: I've asked ChatGPT to simplify it. I should have tested it. Sorry. @kdschlosser please use Liam's suggestion as it's tested and simpler.

**Comment by kisvegabor** [comment]
@liamHowatt or @kdschlosser can you finish this PR based on [this](https://github.com/lvgl/lvgl/pull/6961#discussion_r1832270823)?

**Review by kisvegabor — APPROVED** [review]
Thank you! :heart:

**Review by liamHowatt — APPROVED** [review]

**Pull request merged**